src/data/polynomial/cyclotomic_polynomials.lean

```diff
@@ -0,0 +1,335 @@
+polynomial with integer coefficients. -/
+lemma int_coeff_of_cycl {ζ : K} {n : ℕ} (h : is_primitive_root ζ n) :
+  (∃ (P : polynomial ℤ), map (int.cast_ring_hom K) P = cyclotomic' n K ∧
+  P.degree = (cyclotomic' n K).degree ∧ P.monic) :=
+begin
+  refine lifts_and_degree_eq_and_monic _ (cyclotomic'.monic n K),
+  revert h ζ,
+  apply nat.strong_induction_on n,
+  intros k hk z hzeta,
+  cases nat.eq_zero_or_pos k with hzero hpos,
+  { use 1,
+    simp only [hzero, cyclotomic'_zero, set.mem_univ, subsemiring.coe_top, eq_self_iff_true,
+    map_one, ring_hom.coe_of, and_self], },
+  by_cases hone : k = 1,
+  { use X - 1,
+    simp only [hone, cyclotomic'_one K, set.mem_univ, pnat.one_coe, subsemiring.coe_top,
+    eq_self_iff_true, map_X, map_one, ring_hom.coe_of, and_self, map_sub] },
+  let B : polynomial K := ∏ i in nat.proper_divisors k, cyclotomic' i K,
+  have Bmo : B.monic,
+  { apply monic_prod_of_monic,
+    intros i hi,
+    exact (cyclotomic'.monic i K) },
+  have Bint : B ∈ lifts (int.cast_ring_hom K),
+  { refine subsemiring.prod_mem (lifts (int.cast_ring_hom K)) _,
+    intros x hx,
+    have xsmall := (nat.mem_proper_divisors.1 hx).2,
+    obtain ⟨d, hd⟩ := (nat.mem_proper_divisors.1 hx).1,
+    rw [mul_comm] at hd,
+  exact hk x xsmall (is_primitive_root.pow hpos hzeta hd) },
+  replace Bint := lifts_and_degree_eq_and_monic Bint Bmo,
+  obtain ⟨B₁, hB₁, hB₁deg, hB₁mo⟩ := Bint,
+  let Q₁ : polynomial ℤ := (X ^ k - 1) /ₘ B₁,
+  have huniq : 0 + B * cyclotomic' k K = X ^ k - 1 ∧ (0 : polynomial K).degree < B.degree,
+  { split,
+    { rw [zero_add, mul_comm, ← C_1, (X_pow_sub_one_eq_prod_cycl' hpos hzeta),
+      nat.divisors_eq_proper_divisors_insert_self_of_pos hpos],
+      simp only [true_and, finset.prod_insert, not_lt, nat.mem_proper_divisors, dvd_refl] },
+    rw [degree_zero, bot_lt_iff_ne_bot],
+    intro habs,
+    exact (monic.ne_zero Bmo) (degree_eq_bot.1 habs) },
+  replace huniq := div_mod_by_monic_unique (cyclotomic' k K) (0 : polynomial K) Bmo huniq,
+  simp only [lifts, ring_hom.coe_of, ring_hom.mem_srange],
+  use Q₁,
+  rw [(map_div_by_monic (int.cast_ring_hom K) hB₁mo), hB₁, ← huniq.1],
+  simp
+end
+
+/-- If `K` is of characteristic `0` and there is a primitive `n`-th root of unity in `K`,
+then `cyclotomic n K` comes from a unique polynomial with integer coefficients. -/
+lemma unique_int_coeff_of_cycl [char_zero K] {ζ : K} {n : ℕ+} (h : is_primitive_root ζ n) :
+  (∃! (P : polynomial ℤ), map (int.cast_ring_hom K) P = cyclotomic' n K) :=
+begin
+  obtain ⟨P, hP⟩ := int_coeff_of_cycl h,
+  rw exists_unique,
+  use [P, hP.1],
+  intros Q hQ,
+  have mapinj : function.injective (map (int.cast_ring_hom K)),
+  { apply map_injective,
+    simp only [int.cast_injective, int.coe_cast_ring_hom] },
+  rw [function.injective] at mapinj,
+  apply mapinj,
+  rw [hP.1, hQ]
+end
+
+end field
+
+section integers
+
+/-- The `n`-th cyclotomic polynomial with coefficients in `ℤ`. -/
+def cyclotomic : ℕ → polynomial ℤ
```

Comment:
I would prefer if this definition takes a coefficient ring as argument: `cyclotomic n R`,
and then include a simp-lemma that says: for any `f : R ->+* S` we have `map f (cyclotomic n R) = cyclotomic n S`.

Reply:
OK! In this case it is also reasonable to add that `cyclotomic n R = cyclotomic' n R` in the presence of the relevant primitive roots of unity. This can be proved by induction using the product formula... does someone see a better proof?

Reply:
It probably makes sense to add that lemma, yes.

Reply:
It would make this PR even longer. So it might make sense to PR the changes to the old files in a separate PR.

Reply:
OK, I know how to do it!